src/main/scala/mesosphere/marathon/storage/migration/Helper.scala

```diff
@@ -176,3 +181,105 @@ object InstanceMigration extends MaybeStore with StrictLogging {
 
   def appTask(tasksMap: Map[Task.Id, Task]): Option[Task] = tasksMap.headOption.map(_._2)
 }
+
+object ServiceMigration extends StrictLogging with MaybeStore {
+  /**
+    * Helper method to migrate all versions for all apps by loading them, threading them through the provided flow, and
+    * then storing them. Filtered entities are simply unaffected in persistence.
+    *
+    * @param storageVersion The storage version of the migration being run
+    * @param persistenceStore Used to load and store the entities
+    * @param migratingFlow The flow which actually performs the migrations.
+    * @return
+    */
+  def migrateAppVersions(storageVersion: StorageVersion, persistenceStore: PersistenceStore[ZkId, String, ZkSerialized],
+    migratingFlow: Flow[(Protos.ServiceDefinition, Option[OffsetDateTime]), (Protos.ServiceDefinition, Option[OffsetDateTime]), NotUsed])(implicit ec: ExecutionContext, mat: Materializer): Future[Done] = {
+    implicit val appProtosUnmarshaller: Unmarshaller[ZkSerialized, Protos.ServiceDefinition] =
+      Unmarshaller.strict {
+        case ZkSerialized(byteString) => Protos.ServiceDefinition.parseFrom(byteString.toArray)
+      }
+
+    implicit val appProtosMarshaller: Marshaller[Protos.ServiceDefinition, ZkSerialized] =
+      Marshaller.opaque(appProtos => ZkSerialized(ByteString(appProtos.toByteArray)))
+
+    implicit val appIdResolver: IdResolver[PathId, Protos.ServiceDefinition, String, ZkId] =
+      new ZkStoreSerialization.ZkPathIdResolver[Protos.ServiceDefinition]("apps", true, AppDefinition.versionInfoFrom(_).version.toOffsetDateTime)
+
+    val countingSink: Sink[Done, NotUsed] = Sink.fold[Int, Done](0) { case (count, Done) => count + 1 }
+      .mapMaterializedValue { f =>
+        f.map(i => logger.info(s"$i apps modified when migrating to ${storageVersion}"))
+        NotUsed
+      }
+
+    maybeStore(persistenceStore).map{ zkStore =>
+      zkStore
+        .ids()
+        .flatMapConcat(appId => zkStore.versions(appId).map(v => (appId, Some(v))) ++ Source.single((appId, Option.empty[OffsetDateTime])))
+        .mapAsync(Migration.maxConcurrency) {
+          case (appId, Some(version)) => zkStore.get(appId, version).map(app => (app, Some(version)))
+          case (appId, None) => zkStore.get(appId).map(app => (app, Option.empty[OffsetDateTime]))
+        }
+        .collect{ case (Some(appProtos), optVersion) if !appProtos.hasRole => (appProtos, optVersion) }
```

Comment:
This is thought to be a general migration flow, correct? Why would we filter out apps without roles then? I understand that this is an invalid case but so are other app fields too.